toptal/picasso
# feat: add codemod

[FX-4260]

### Description

This pull request adds a codemod that replaces BASE-aligned spacing property values with new spacing constants.

`jscodeshift` has the following problem when running this codemod:
- empty line is inserted before the import that is added via `insertAfter()`, more in https://github.com/facebook/jscodeshift/issues/185
- the semicolon is inserted after the import that is added via `insertAfter()`, more in https://github.com/benjamn/recast/issues/226

However, both issues are solved by running lint or prettier after the codemod run (this recommendation was added to the changeset)

### How to test

- Run tests – check out this branch locally and run `yarn run jest packages/picasso-codemod/src/v38.1.0/__tests__`
- Run actual codemod in Staff Portal – check out https://github.com/toptal/staff-portal/pull/10942 branch locally, drop last commit, and run `npx @toptal/picasso-codemod v38.1.0`

### Screenshots

Typical change from https://github.com/toptal/staff-portal/pull/10942/files

<img width="496" alt="Screenshot 2023-09-01 at 12 37 38" src="https://github.com/toptal/picasso/assets/1390758/ad5ce5e1-79ab-48e3-aca6-a6ca1fe52576">

---

Running in Staff Portal (https://github.com/toptal/staff-portal/pull/10942)

<img width="1241" alt="Screenshot 2023-09-01 at 09 18 53" src="https://github.com/toptal/picasso/assets/1390758/f3d5d4a6-f531-4995-85ea-ee1b990a797f">

### Development checks

- [N/A] Make sure that additions and changes on the design follow [Toptal's BASE design](https://design.toptal.net/), and it's been already discussed with designers at #-base-core
- [N/A] Annotate all `props` in component with documentation
- [N/A] Create `examples` for component
- [N/A] Ensure that deployed demo has expected results and good examples
- [N/A] Ensure the changed/created components have not caused accessibility issues. [How to use accessibility plugin in storybook](https://github.com/toptal/picasso/blob/master/docs/contribution/accessibility.md).

> All **_development checks_** should be done and set checked to pass the
> **GitHub Bot: TODOLess** action

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run package:alpha-release` - Release alpha version
- `@toptal-anvil ping reviewers` - Ping FX team for review

</details>

<details>
<summary>PR Review Guidelines</summary>
<br />

#### When to approve? ✅

**You are OK** with merging this PR and

1. You have no extra requests.
2. You have optional requests.
   1. Add `nit:` to your comment. (ex. `nit: I'd rename this variable from makeCircle to getCircle`)

#### When to request changes? ❌

**You are not OK** with merging this PR because

1. Something is broken after the changes.
2. Acceptance criteria is not reached.
3. Code is dirty.

#### When to comment (neither ✅ nor ❌)

**You want your comments to be addressed** before merging this PR in cases like:

1. There are leftovers like unnecessary logs, comments, etc.
2. You have an opinionated comment regarding the code that requires a discussion.
3. You have questions.

#### How to handle the comments?

1. An owner of a comment is the only one who can resolve it.
2. An owner of a comment must resolve it when it's addressed.
3. A PR owner must reply with ✅ when a comment is addressed.

</details>


[FX-4260]: https://toptal-core.atlassian.net/browse/FX-4260?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
